### ISSUE TITLE
Automatically pre-pull kernel images

### ISSUE THREAD

**Issue by witten** [jupyter/enterprise_gateway#638]
Based on the comments below, this ticket has morphed into a feature to pre-pull kernel images to all K8s nodes, so as to sidestep Jupyter's kernel start timeout. This includes pre-pulling:

 * New kernel images, e.g. on a fresh deploy of Enterprise Gateway to a cluster or the addition of a new kernelspec to an existing deployment.
 * Updated kernel images, e.g. a given image name/tag doesn't change, but its underlying image has changed due to a new release.
 * New K8s nodes being added, e.g. due to auto-scaling.

Ideally this feature would dynamically source the set of images to pre-pull from the configured kernelspecs.

#### Original Description

I'm trying to make a custom Spark kernel as per [this comment](https://github.com/jupyter/enterprise_gateway/issues/625#issuecomment-479698455). I've set up a custom kernelspec as per [these docs](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/docker.html#deploying-your-custom-kernel-image) and mounted into the Enterprise Gateway pod [via a custom container image](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/getting-started-kubernetes.html#via-custom-container-image). That all appears to work fine.

However, when the kernel launches a driver pod, it does so with an `imagePullPolicy` of `IfNotPresent` (the default), apparently because one is [not otherwise configured](https://github.com/jupyter/enterprise_gateway/blob/master/etc/kernel-launchers/kubernetes/scripts/kernel-pod.yaml.j2#L55). I'd like to be able to set a different `imagePullPolicy`, e.g. `Always`, because now I'm in a position where I'd like to rev my driver image, but Enterprise Gateway refuses to pull updates and I have no way to override that behavior!

So this is a feature request to provide *some* mechanism for setting the `imagePullPolicy` for the kernel launcher, or by some other mechanism to allow my driver image to receive updates in the cluster.

Relevant section of kernelspec:

```json
  "metadata": {
    "process_proxy": {
      "class_name": "enterprise_gateway.services.processproxies.k8s.KubernetesProcessProxy",
      "config": {
        "image_name": ".../jupyter-python-spark-24-job:initial-import",
        "executor_image_name": ".../jupyter-python-spark-24-job:initial-import"
      }
    }
  },
```

Needless to say, the image tag is not changing with each revision to the image. (If it did, this wouldn't be an issue.)

And here's evidence of the resulting `imagePullPolicy`:

```bash
$ kubectl get pod -n guest-2cd0b88c-... --show-all -o yaml |less
```

```yaml
...
 spec:
    containers:
...
      image: .../jupyter-python-spark-24-job:initial-import
      imagePullPolicy: IfNotPresent
...
```

**Comment by kevin-bates** [comment]
Yeah, I think this was `Always` for a brief moment, but I found this difficult for development and figured `IfNotPresent` was sufficient.  I'm now thinking that it should probably be changed since `Always` will pull from the cache if it's already there.  I think it's crucial for kernel-based images to be pre-populated, since we can't incur the image download times as a portion of the kernel startup (which is essentially on a timer).  So that's the concern. 

When the image is not present, these two policies behave the same.  If present, is `Always` fast enough to not interfere with startup times?  That's probably worth an exercise.

**Comment by witten** [comment]
Some brainstorming on the startup time issue: Rather than requiring the user to pre-pull the image, would it be possible to somehow programmatically pre-pull it? I don't know when the Jupyter timer starts ticking, but one idea is to add an `initContainer` to the [kernel pod](https://github.com/jupyter/enterprise_gateway/blob/master/etc/kernel-launchers/kubernetes/scripts/kernel-pod.yaml.j2) that does nothing but spin up the kernel image and exit (thereby implicitly pre-pulling it). If that's too late for the timer, maybe there is an earlier point where an explicit pull can happen automatically?

**Comment by esevan** [comment]
I agree warming up the nodes with the kernel images is required for EG operation, and `imagePullPolicy` is meaningless when all nodes are warmed with the latest kernel images;
Otherwise, the users will be mad at the service.. too long kernel connection time :(...
+ User's clock starts ticking right after selecting the kernel. EG starts ticking after sending pod creation request to k8s. It includes (1. `initContainer` is created, 2. `kernel` container is up and 3. `kernel` connects to EG)

It is like another issue how to pre-pull the kernel images into the nodes.
I think it's hard to expect when the kernel image should be pre-pulled because it depends on user behavior. What about using independent service to pre-pull the images? At this moment, I'm using `daemonSet` to pre-pull the images to the nodes [reference](https://gist.github.com/esevan/62e4773c405be3cfc30bc527330d18c4). I think it can be extended to the service handling pre-pulling request.

**Comment by kevin-bates** [comment]
Right.  The timer starts when the request to start a kernel is received.  We've talked about warm starts into a pool, but that's a slippery slope when it comes to per-user/request resources, etc.

I'm wondering if we could tie something like this into the kernelspec manager that looks for image entries in the available kernelspecs and pulls then.  Of course, that would only populate the images on the EG node.

I thought I recalled some kind of pre-pull policy thing that's more scoped at k8s.  I wonder if there's some way we could tie that kind of thing into the kernelspec manager.  Then, when EG starts we could force a hit on the kernespec manager.

**Comment by kevin-bates** [comment]
@esevan - Our responses came in about the same time.  I like your idea and wonder if we should explore that with kernelspec manager - since that's what would gather up the applicable specs (to determine applicable images).

**Comment by esevan** [comment]
@kevin-bates it would be great if EG tracks pre-pulling process of kernel described in kernelspecs and returns available kernels to NB2KG rather than just returing whitelisted kernels. I'm wondering what trigger makes EG pulls the images. It is simple when EG first starts, but complicated image is changed in docker registry. I don't think it is good practice to manage the kernel version, but should we cover that case?

+ If we had admin dashboard, I think we could handle that case as well by triggering image update via dashboard.

**Comment by kevin-bates** [comment]
>I'm wondering what trigger makes EG pulls the images. It is simple when EG first starts, but complicated when the image with same tag is changed in docker registry. I don't think it is good practice to manage the kernel version, but should we cover that case?

Right. My last response was going to contain a comment about this being a slippery slope in that we don't want to go overboard. 

I'm thinking we should assess the cost of `Always` when the image is present.  If the startups are within the acceptable range (i.e., a second or two longer than if we used `IfNotPresent`) then I think we're okay for now.

This sounds like an awesome feature for an admin dashboard app!

Btw, the whitelist is purely for example purposes.  We could remove the property and then ALL kernelspecs are returned.   Now that we have "platform-specific" tar files of kernelspecs, we could probably rely on those and remove the whitelist.  This way adding kernelspecs wouldn't require EG redeployment/restarts.

**Comment by witten** [comment]
A lot of interesting ideas here.. I had no idea this was such a thorny problem. The DaemonSet idea and/or dedicated pre-pulling service is kind of clever. Replying to a few individual points:

> When the image is not present, these two policies behave the same. If present, is Always fast enough to not interfere with startup times? That's probably worth an exercise.

Yeah. I would imagine it largely depends on the delta in the before-and-after images. For instance, if it's a minor change and only a single layer updates, it probably would not interfere with startup times. On the other hand, if it's a major update and, say, the whole base image changes from Ubuntu to Alpine or the other way around, it would have to re-pull all layers, and a timeout would be quite likely to occur.

> User's clock starts ticking right after selecting the kernel. EG starts ticking after sending pod creation request to k8s. It includes (1. initContainer is created, 2. kernel container is up and 3. kernel connects to EG)

Ugh, okay. That certainly makes it more difficult.

> I thought I recalled some kind of pre-pull policy thing that's more scoped at k8s.

You might be thinking of an `AlwaysPullImages` [admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#what-does-each-admission-controller-do)? If so, it effectively just sets `imagePullPolicy: Always` on everything, which may solve the initial problem that triggered this ticket, but doesn't solve the image pre-pulling / timeout problem.

> It is simple when EG first starts, but complicated when the image with same tag is changed in docker registry. I don't think it is good practice to manage the kernel version, but should we cover that case?

Riffing on a couple of the ideas mentioned here: When Enterprise Gateway starts, it could "publish" a list of its configured kernelspecs' images somewhere (e.g. a ConfigMap). Then, a DaemonSet like the one linked above could read that list of images and pull each one. And to handle the "same tag is changed in the Docker registry" case, that Daemonset could even re-pull the images periodically in case there are changes. However, that might need to be paired with [other work-arounds](https://github.com/kubernetes/kubernetes/issues/33664) to also kick the Pod/Deployment/whatever to make use of any newly pulled image.

Having said all that, I wonder why something like this doesn't already existing for Kubernetes in general? I know there are [solutions for plain Docker](https://github.com/containrrr/watchtower), for instance. (Maybe Watchtower can just be shoved into a DaemonSet to handle this.)

**Comment by esevan** [comment]
Watchtower sounds great! I thought we could re-pull images only by periodic job. If there's a watchover function on kernelspec manager in EG or other microservice, it would ease off the repetitive pain pulling image after each update.

Let's assume we have that kind of function. Watchtower re-pulls kernel images in registry update. I think it is OK to perform blue-green kernel deployment which means EG just selects old kernel while update is in progress by selecting `ifNotPresent` policy and selects the new kernel after update is complete. If this scenario is reasonable it would be good to maintain current `ifNotPresent` policy and make Watchtower to be in charge of kernel update.

**Comment by lresende** [comment]
JupyterHub has the concept of [image puller](https://zero-to-jupyterhub.readthedocs.io/en/latest/optimization.html?highlight=puller) and it looks like just what is being discussed here. It can be configured to pre-pull images, but I also believe it has the capabilities to watch for any new/updated ones.

**Comment by witten** [comment]
> If this scenario is reasonable it would be good to maintain current ifNotPresent policy and make Watchtower to be in charge of kernel update.

Sounds good. Shall I file a new ticket for Watchtower / JupyterHub's image puller, or just repurpose this one?

**Comment by witten** [comment]
FYI here is the business end of the JupyterHub [continuous image puller DaemonSet](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/master/jupyterhub/templates/image-puller/_daemonset-helper.yaml). It doesn't need to do an explicit `docker pull` because it declares the images to manage as initContainers that immediately exit, which implicitly pulls them. While this does cause managed images to get pulled to newly added K8s nodes, I don't see anything here that would cause an existing image to get updated automatically. (I may just be missing it.)

**Comment by lresende** [comment]
With my suggestion, I just wanted to point how hub looked into this problem from their perspective. 

For Enterprise Gateway, we could use JupyterHub's solution, do something based on what they have done and adapt to our needs, build something specific to EG.

As for the ticket, if this image puller concept will resolve your needs, we can use this ticket and clarify the title if needed.

**Comment by kevin-bates** [comment]
Great discussion!  

I wanted to circle back to @witten's original issue regarding `imagePullPolicy` in the context of Spark-based images.  Had this been a vanilla k8s kernel, you'd just adjust the imagePullPolicy attribute of the kernel-pod jinja template that's part of the kernelspecs file hierarchy and my original response was (incorrectly) along these lines.  However, because k8s spark does not yet support custom pod templates (I keep forgetting that spark on k8s launches are not really under our control), you must perform any "customization" via a limited set of supported parameters.  In this case, they do support [`spark.kubernetes.container.image.pullPolicy`](https://spark.apache.org/docs/latest/running-on-kubernetes.html#spark-properties) and you would add that to your kernelspec's `SPARK_OPTS` env.

Regarding the rest of the discussion..., 
Looking at the DaemonSet stuff, would it make sense to create a simple image that periodically hits the EG /api/kernelspecs endpoint and either checks for the image's existence or unconditionally pulls the image based on a command-line option?  This would probably address initial starts and additional nodes.  Where it might still be an issue is when a new kernelspec gets added, and the interval used in the puller container hasn't fired before the first user requests that kernel.  However, we could document that admins creating new kernelspecs that reference new images should either wait for an interval or pull the image themselves before attempting kernel usage.  There might be a way to decrease this window via k8s watching - but I think we'd need to subclass the KernelSpecManager in order to detect new images and notify the DaemonSet puller "service".  Perhaps that could be implemented once a poll-based POC has proven worthy.

**Comment by witten** [comment]
K8s Spark does support custom pod templates (at least in master)! Details at https://github.com/apache/spark/blob/master/docs/running-on-kubernetes.md#pod-template

**Comment by kevin-bates** [comment]
Yep - that's what we're waiting on: #559

**Comment by kevin-bates** [comment]
Hello - I spent a few hours thinking and implementing some basic puller logic.  I thought we could talk about it so I've left the PR as a WIP.  I've built a yaml file and have it running as a DaemonSet, but wanted to convert the yaml to a helm template as part of the merge.

FWIW: here's the initial (non-templated) yaml.  In addition, I thought this would prevent its running on the master node, but it didn't...
```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: kernel-image-puller
  namespace: enterprise-gateway
spec:
  selector:
    matchLabels:
      name: kernel-image-puller 
  template:
    metadata:
      labels:
        name: kernel-image-puller 
        app: enterprise-gateway
        component: kernel-image-puller
    spec:
      tolerations:
      - key: node-role.kubernetes.io/master
        effect: NoSchedule
      containers:
      - name: kernel-image-puller 
        image: elyra/kernel-image-puller:dev
        env:
          - name: KIP_GATEWAY_HOST
            value: 'http://k8s-omg-node-1.fyre.ibm.com/gateway'
        volumeMounts:
          - name: dockersock
            mountPath: "/var/run/docker.sock"
      volumes:
      - name: dockersock
        hostPath:
          path: /var/run/docker.sock
```

Here's the output from one of the nodes...
```
[I 2019-04-17 01:29:38,401 kernel_image_puller.t1] Pulled image 'elyra/kernel-tf-py:dev' in 1.640449047088623 secs.
[I 2019-04-17 01:29:38,401 kernel_image_puller.t1] Pulling image 'elyra/kernel-py:dev'...
[I 2019-04-17 01:29:38,423 kernel_image_puller.t2] Pulled image 'elyra/kernel-tf-gpu-py:dev' in 1.659637451171875 secs.
[I 2019-04-17 01:29:38,424 kernel_image_puller.t2] Pulling image 'elyra/kernel-scala:dev'...
[I 2019-04-17 01:29:40,031 kernel_image_puller.t1] Pulled image 'elyra/kernel-py:dev' in 1.6293447017669678 secs.
[I 2019-04-17 01:29:40,031 kernel_image_puller.t1] Pulling image 'kbates/kernel-spark-r:workdir'...
[I 2019-04-17 01:29:40,109 kernel_image_puller.t2] Pulled image 'elyra/kernel-scala:dev' in 1.6853554248809814 secs.
[I 2019-04-17 01:29:40,109 kernel_image_puller.t2] Pulling image 'kbates/kernel-scala:workdir'...
[I 2019-04-17 01:29:41,691 kernel_image_puller.t1] Pulled image 'kbates/kernel-spark-r:workdir' in 1.6596968173980713 secs.
[I 2019-04-17 01:29:41,691 kernel_image_puller.t1] Pulling image 'elyra/kernel-r:dev'...
[I 2019-04-17 01:29:41,765 kernel_image_puller.t2] Pulled image 'kbates/kernel-scala:workdir' in 1.6558308601379395 secs.
[I 2019-04-17 01:29:41,766 kernel_image_puller.t2] Pulling image 'kbates/kernel-spark-py:workdir'...
[I 2019-04-17 01:29:43,333 kernel_image_puller.t1] Pulled image 'elyra/kernel-r:dev' in 1.642183542251587 secs.
[I 2019-04-17 01:29:43,609 kernel_image_puller.t2] Pulled image 'kbates/kernel-spark-py:workdir' in 1.8437349796295166 secs.
[I 2019-04-17 01:29:43,610 kernel_image_puller.MainThread] Images pulled.  Sleeping 300 seconds...
[I 2019-04-17 01:34:43,712 kernel_image_puller.MainThread] Fetching kernelspecs from 'http://k8s-omg-node-1.fyre.ibm.com/gateway/api/kernelspecs' ...
[I 2019-04-17 01:34:43,746 kernel_image_puller.t2] Pulling image 'elyra/kernel-tf-py:dev'...
[I 2019-04-17 01:34:43,749 kernel_image_puller.t1] Pulling image 'elyra/kernel-tf-gpu-py:dev'...
[I 2019-04-17 01:34:45,429 kernel_image_puller.t1] Pulled image 'elyra/kernel-tf-gpu-py:dev' in 1.679419755935669 secs.
[I 2019-04-17 01:34:45,429 kernel_image_puller.t1] Pulling image 'elyra/kernel-py:dev'...
[I 2019-04-17 01:34:45,498 kernel_image_puller.t2] Pulled image 'elyra/kernel-tf-py:dev' in 1.7519581317901611 secs.
[I 2019-04-17 01:34:45,498 kernel_image_puller.t2] Pulling image 'elyra/kernel-scala:dev'...
[I 2019-04-17 01:34:47,074 kernel_image_puller.t1] Pulled image 'elyra/kernel-py:dev' in 1.644660234451294 secs.
[I 2019-04-17 01:34:47,074 kernel_image_puller.t1] Pulling image 'kbates/kernel-spark-r:workdir'...
[I 2019-04-17 01:34:47,151 kernel_image_puller.t2] Pulled image 'elyra/kernel-scala:dev' in 1.653134822845459 secs.
[I 2019-04-17 01:34:47,151 kernel_image_puller.t2] Pulling image 'kbates/kernel-scala:workdir'...
[I 2019-04-17 01:34:48,733 kernel_image_puller.t1] Pulled image 'kbates/kernel-spark-r:workdir' in 1.6592676639556885 secs.
[I 2019-04-17 01:34:48,734 kernel_image_puller.t1] Pulling image 'elyra/kernel-r:dev'...
[I 2019-04-17 01:34:48,910 kernel_image_puller.t2] Pulled image 'kbates/kernel-scala:workdir' in 1.758805274963379 secs.
[I 2019-04-17 01:34:48,911 kernel_image_puller.t2] Pulling image 'kbates/kernel-spark-py:workdir'...
[I 2019-04-17 01:34:50,446 kernel_image_puller.t1] Pulled image 'elyra/kernel-r:dev' in 1.7126076221466064 secs.
[I 2019-04-17 01:34:50,706 kernel_image_puller.t2] Pulled image 'kbates/kernel-spark-py:workdir' in 1.7953824996948242 secs.
```

Comments, questions and general discussion are welcome!

**Comment by esevan** [comment]
What a quick response with source code! Respect!

> I thought this would prevent its running on the master node, but it didn't...

I think `toleration = {"key": "node-role.kubernetes.io/master", "effect": "NoSchedule"}` enables scheduling on master node.

**Comment by kevin-bates** [comment]
Awesome - I admit I didn't take time to understand that stanza, but glad to hear this!

I think when converting this to helm (daemonset.yaml ?), we're going to need to add some ingress-related variables to the Values.yaml file in order to "compose" `KIP_GATEWAY_HOST`.

**Comment by witten** [comment]
This is super cool. If the kernelspecs endpoint is running on the same cluster as the image puller, which I assume is the case, then why do you need a gateway hostname or Ingress for it? Wouldn't you be able to just use the internal Kubernetes DNS + routing?

**Comment by kevin-bates** [comment]
That's a correct assumption.  I had totally forgot about that kind of thing.  Are you referring to using something like this:
```yaml
        env:
          - name: KIP_GATEWAY_HOST
            value: 'http://enterprise-gateway.enterprise-gateway:8888'
```
because I just updated the yaml above and it works like a charm!  
If you're instead referring to this kind of thing: `enterprise-gateway.enterprise-gateway.svc.cluster.local`, I couldn't get that work (brief attempt using `http://` as a prefix).

Question (assuming the service.namespace:port format is the correct approach):
When migrating to helm should this get decomposed into helm values?  Looks like most of these would come from [`service.yaml`](https://github.com/jupyter/enterprise_gateway/blob/master/etc/kubernetes/helm/templates/service.yaml).  Is it best to make variables of the service `name`, and `ports.name`, `ports.port` and `ports.targetPort`?

**Comment by witten** [comment]
Broadly speaking, yes, I was referring to a hostname like one of the ones you used. Whichever one works.

As to your question, I think it would be keeping with Helm convention to make the gateway host configurable as one or more Helm value.  However, given that `service.yaml` has namespace, pod name, and port all hard-coded, you don't strictly have to make the gateway host configurable right now. You could just hard-code it into the Helm template until such a time as it's also configurable in `service.yaml`.

**Comment by kevin-bates** [comment]
Yeah, I noticed that `{{ .Release.Namespace }}` is used elsewhere and was just going to fix that up in `service.yaml` now.  Seems like that was just missed - unless I'm missing something.  I'm guessing `{{ .Release.Namespace }}` is the value pulled from the `--namespace` parameter on the `helm` command - is that correct?

**Comment by witten** [comment]
You're right, good catch! In that case, it could make sense to have one or a couple of Helm values for the gateway host.

**Comment by esevan** [comment]
> > "hold off using a given kernel until the cluster has had a chance to pull"
>
> Isn't this good to avoid "timeout" issue? Looks like "rolling upgrade" and "zero-downtime" upgrade.

I thought of this last night. Too long time distance with community T^T...
It is great to start as it is (Thank you so much for this contribution!), but I think concept of "Pulling images which are written in `kernelspecs`" can be improved in such manner in that "Pulling image first and update `kernelspecs`".

In my exp, customized kernel update is with following process.
1. Push a new image to Harbor (docker repo)
1. Pull the image in all EG nodes.
1. Update kernelspec image name so that users are able to use kernel without timeout.

Present contribution is switching step 2 and step3. What do you think of above process folks?

**Comment by kevin-bates** [comment]
Yes - I agree that switching steps 2 and 3 is preferred and the ideal process - I'm just not sure how'd we get there.  I'd rather not have to maintain a separate list of kernel image names - so retaining the context relative to the kernelspec still strikes me as a requirement.  Some ideas...

1. I think introducing the notion of something like a _kernelspec editor_ may be one way to somehow convey the fact we want to seed nodes with a future kernel image.  
1. Another idea would be to perhaps introduce a kernelspec _staging area_ where the puller may hit a REST API that includes that area - although I'm not sure when an admin would know its okay to move the kernelspec out of _staging_ and into the main area.
1. Another way to perhaps _"seed before use"_ is use of the white list.  If the puller could say "give me ALL kernelspecs" (not just whitelisted specs), then the kernel image could be propagated and only after nodes are seeded (again not sure how that is confirmed), then the new kernelspec is added to the whitelist - where its available to the general population.

    \<rambling thoughts>
    Btw, the whitelist is problematic in that it requires an EG restart to pick up the new values.
 
    Hmm - I think the config traitlets in jupyter allow the ability the detect a change in value (via [`@observe`](https://github.com/ipython/traitlets/blob/master/traitlets/traitlets.py#L813-L834)).  This implies that if we use the configuration file to maintain the whitelist (rather than env), then we might be able to "see" its value has changed.  This might be something to look into.  I also wonder if that config file could be placed into a ConfigMap, or some k8s object, that can then be more easily modified.  If we can solve the whitelist update issue, then I suspect we'd be good to go.  Then it's a matter of having a REST endpoint (or parameter to the existing endpoint) that indicates the whitelist should be ignored so the puller sees _"before use"_ kernelspecs and no "move from staging" is required.
    </rambling thoughts>

I think the last item is probably worth looking into and would benefit all platforms in eliminating the need to restart for new kernelspecs yet retain the notion of not making a kernelspec immediately available by virtue of its existence.

**Comment by kevin-bates** [comment]
Bummer - my hunch on `observe` is totally wrong.  It only exists to detect changes to the traitlets _within_ the application.  As a result, something would need to hit the EG instance and update its `KernelSpecManager.whitelist` variable.  I don't think `@observe` is even useful for our use case.

However, given that the whitelist is consulted on each fetch of kernelspecs, it is relatively dynamic.  As a result, I suppose we could introduce a _"dynamic whitelist"_ where there is no whitelist configured, then the initial load of kernelspecs _pins_ them into a whitelist.  This way, new kernelspecs could be added but not get used until the "hit the EG instance and update its `KernelSpecManager.whitelist` variable" request happens (which doesn't take place until KIP has updated the nodes).

Hmm - if EG is restarted before the "hidden" kernelspecs are added to the whitelist, they'd get included in the dynamic whitelist upon restart - so this is sketchy at best.

**Comment by witten** [comment]
>>        "hold off using a given kernel until the cluster has had a chance to pull"
>
>    Isn't this good to avoid "timeout" issue? Looks like "rolling upgrade" and "zero-downtime" upgrade.

> In my exp, customized kernel update is with following process.
>
>    1. Push a new image to Harbor (docker repo)
>    2. Pull the image in all EG nodes.
>    3. Update kernelspec image name.
>
> By doing above process, users are able to use kernel without timeout.

I may be missing something, but our usual process is a little different and perhaps makes this simpler:

1. Push a new image to our Docker container registry and tag it with the same image tag that's already in use in the cluster. E.g., ":latest" or similar. Could even be a major version like ":3". It could also get tagged with a more specific tag like ":3.0.2", but that's not the tag we're referring to in the kernelspec.
2. Re-pull the image on all EG nodes, using the same tag that's already in place.
3. That's it! No need to update kernelspecs, and there's no intervening period where timeouts happen. Any given kernel gets either the original image or, as soon as it's done pulling, the new image.

(Note that timeouts can still happen on newly created K8s nodes before the initial pull happens.)

**Comment by kevin-bates** [comment]
Yes - retaining the image name/tag of in-use images is an approach that relieves a lot of issues.  Then one only needs to set `KIP_PULL_POLICY=Always` and they're good to go ... until they introduce a new image to the mix.  

If they have their own means of pulling images to nodes, sweet, but we're trying to eliminate the need for a) maintaining a separate list of image names outside those scattered across the kernelspecs and b) having to pull those images in a manner so as to not incur first-use timeouts.

Hmm,  I suppose a site could run two EG instances, probably in two different namespaces, one for staging the other for prod.  They could validate new and updated kernelspecs in staging - where a different whitelist is in use and the staging KIP pulls the new images.  (In staging, "users" would need to know that launching a kernel prior to its "distribution" may timeout.)  Then, once ready, they update the whiltelist of prod and redeploy - where the images will already be present.

(btw, redeployment is an issue that I'm hoping is solved with kernel-session persistence to shared storage.  Then, the newly started EG can re-establish connections to existing kernels - but this is a different topic. :smile:)

**Comment by witten** [comment]
Got it. So you're talking about the "roll out a completely new image with a new name/tag" use case. A variant idea for that: Run a single EG instance, and add the new image to the whitelist as you normally would. But then EG could omit returning image names to users that KIP has not yet pulled on all nodes. Either by querying KIP for that information or somehow divining it by itself. Not sure if that approach works.

**Comment by kevin-bates** [comment]
Yes, that's the use case.  Right, the problem is that KIP doesn't know when its done across all nodes.  It can only determine the node on which it's running.  Also, the update of the whitelist itself forces a restart so I suspect 24/7 sites would forgo the whiltelist filter altogether.

Again, this requires override of KernelSpecManager, but I wonder if we could add a piece of metadata into kernel.json that indicates a given kernel is "enabled" and have the query issued by KIP (or future KernelSpec editors) say "give me all kernelspecs", but the default only return enabled specs.  Admins would then flip the the boolean once they've finalized testing, confirmed image deployment, etc.
